lib/utils/proxy/proxy.go

```diff
@@ -56,19 +51,18 @@ func dialWithDeadline(network string, addr string, config *ssh.ClientConfig) (*s
 // dialALPNWithDeadline allows connecting to Teleport in single-port mode. SSH protocol is wrapped into
 // TLS connection where TLS ALPN protocol is set to ProtocolReverseTunnel allowing ALPN Proxy to route the
 // incoming connection to ReverseTunnel proxy service.
-func dialALPNWithDeadline(network string, addr string, config *ssh.ClientConfig, insecure bool) (*ssh.Client, error) {
+func (d directDial) dialALPNWithDeadline(network string, addr string, config *ssh.ClientConfig) (*ssh.Client, error) {
 	dialer := &net.Dialer{
 		Timeout: config.Timeout,
 	}
 	address, err := utils.ParseAddr(addr)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	tlsConn, err := tls.DialWithDialer(dialer, network, addr, &tls.Config{
-		NextProtos:         []string{string(alpncommon.ProtocolReverseTunnel)},
-		InsecureSkipVerify: insecure,
-		ServerName:         address.Host(),
-	})
+	conf := d.tlsConfig.Clone()
+	conf.ServerName = address.Host()
+	conf.InsecureSkipVerify = d.insecure
```

Comment:
Couple of things:

1. This block repeats multiple times in the directDial and proxyDial. Can it be encapsulated into some `getTLSConfig()` method on them?
2. The DialerFromEnvironment contract doesn't actually prevent anyone from creating it without TLS config so it still can potentially be nil and lead to a panic here. Should we explicitly return an error in this case?

```go
func (d directDial) getTLSConfig(addr <type>) (*tls.Config, error) {
  if d.tlsConfig == nil {
    return nil, trace.BadParameter("directDial TLS config was nil")
  }
  tlsConfig := d.tlsConfig.Clone()
  tlsConfig.ServerName = addr.Host()
  tlsConfig.InsecureSkipVerify = d.insecure
  return tlsConfig, nil
}

func (d proxyDial) getTLSConfig(addr <type>) (*tls.Config, error) {
  // same
}
```